File: 1036-rotting-oranges/1036-rotting-oranges.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        int fresh = 0;
        queue<pair<int, int>> q;

        for (int i = 0; i < grid.size(); ++i){
            for (int j = 0; j < grid[0].size(); ++j){
                if (grid[i][j] == 1) ++fresh;
                if (grid[i][j] == 2) q.push({i, j});
            }
        }

        int time = 0;
        vector<pair<int, int>> directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

        while (!q.empty() && fresh > 0){
            int size = q.size();

            for (int i = 0; i < size; ++i){
                auto [r, c] = q.front();
                q.pop();

                for (auto const& [dr, dc] : directions){
                    int nr = r + dr;
                    int nc = c + dc;

                    if (nr < 0 || nr == grid.size() || nc < 0 || nc == grid[0].size() || grid[nr][nc] != 1) continue;

                    grid[nr][nc] = 2;
                    --fresh;

                    q.push({nr, nc});
                }
            }

            ++time;
        }

        return fresh == 0 ? time : -1;
    }
};

File: 1036-rotting-oranges/1036-rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        minutes = 0
        fresh_count = 0
        q = deque()
        rows = len(grid)
        cols = len(grid[0])
        
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 2:
                    q.append([r, c])
                if grid[r][c] == 1:
                    fresh_count += 1
        
        directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]

        while q and fresh_count > 0:
            
            for i in range(len(q)):
                r, c = q.popleft()
                
                for dr, dc in directions:
                    row = dr + r
                    col = dc + c

                    if row < 0 or row == rows or col < 0 or col == cols or grid[row][col] != 1:
                        continue
                    
                    grid[row][col] = 2
                    q.append([row, col])
                    fresh_count -= 1
            minutes += 1
        
        return minutes if fresh_count == 0 else -1

                    
                    